docs/framework/interop/default-marshaling-for-strings.md

```diff
@@ -1,264 +1,267 @@
+
+```vb
+<StructLayout(LayoutKind.Sequential, CharSet := CharSet.Ansi)> _
+Structure StringInfoA
+    <MarshalAs(UnmanagedType.LPStr)> Public f1 As String
+    <MarshalAs(UnmanagedType.ByValTStr, SizeConst := 256)> _
+    Public f2 As String
+End Structure
+
+<StructLayout(LayoutKind.Sequential, CharSet := CharSet.Unicode)> _
+Structure StringInfoW
+    <MarshalAs(UnmanagedType.LPWStr)> Public f1 As String
+    <MarshalAs(UnmanagedType.ByValTStr, SizeConst := 256)> _
+    Public f2 As String
+<MarshalAs(UnmanagedType.BStr)> Public f3 As String
+End Structure
+
+<StructLayout(LayoutKind.Sequential, CharSet := CharSet.Auto)> _
+Structure StringInfoT
+    <MarshalAs(UnmanagedType.LPTStr)> Public f1 As String
+    <MarshalAs(UnmanagedType.ByValTStr, SizeConst := 256)> _
+    Public f2 As String
+End Structure
+```
+
+```csharp
+[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Ansi)]
+struct StringInfoA
+{
+    [MarshalAs(UnmanagedType.LPStr)] public string f1;
+    [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 256)] public string f2;
+}
+
+[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+struct StringInfoW
+{
+    [MarshalAs(UnmanagedType.LPWStr)] public string f1;
+    [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 256)] public string f2;
+    [MarshalAs(UnmanagedType.BStr)] public string f3;
+}
+
+[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Auto)]
+struct StringInfoT
+{
+    [MarshalAs(UnmanagedType.LPTStr)] public string f1;
+    [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 256)] public string f2;
+}
+```
+
+## Fixed-Length String Buffers
+
+In some circumstances, a fixed-length character buffer must be passed into unmanaged code to be manipulated. Simply passing a string does not work in this case because the callee cannot modify the contents of the passed buffer. Even if the string is passed by reference, there is no way to initialize the buffer to a given size.
+
+The solution is to pass a <xref:System.Text.StringBuilder> buffer as the argument instead of a string. A `StringBuilder` can be dereferenced and modified by the callee, provided it does not exceed the capacity of the `StringBuilder`. It can also be initialized to a fixed length. For example, if you initialize a `StringBuilder` buffer to a capacity of `N`, the marshaler provides a buffer of size (`N`+1) characters. The +1 accounts for the fact that the unmanaged string has a null terminator while `StringBuilder` does not.
+
+For example, the Microsoft Win32 API `GetWindowText` function (defined in Windows.h) is a fixed-length character buffer that must be passed into unmanaged code to be manipulated. `LpString` points to a caller-allocated buffer of size `nMaxCount`. The caller is expected to allocate the buffer and set the `nMaxCount` argument to the size of the allocated buffer. The following code shows the `GetWindowText` function declaration as defined in Windows.h.
```

Comment:
`GetWindowText` is not a function. But P/Invoke will probe for function names following the A/W suffix on Windows. Is this covered in docs?

Reply:
Moreover, this function is unavailable in console apps. Could you think of a better candidate?

Reply:
I linked to GetWindowTextW, since that's most likely the function to be called. Probing for function names, and even the use of an ASCII and a wide-character function, used to be covered in the VB 6 interop docs, but I don't think that it's covered in our doc set.

Finally, I can't think of a better candidate for GetWindowText. The problem with most code in console apps that I've seen, as I recall, is that they retrieve a handle to the console window based on its text, then call GetWindowText to get the text, most of which was already used to retrieve the window handle.

Reply:
P/Invoke probing is mentioned in [DllImportAttribute.CharSet Field](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.dllimportattribute.charset#remarks) and covered in .NET Framework [Specifying a Character Set](https://docs.microsoft.com/dotnet/framework/interop/specifying-a-character-set) article. It's another issue whether we should port them to .NET Guide [Charsets and marshalling](https://docs.microsoft.com/dotnet/standard/native-interop/charset) article in a "Windows Specific" section, or add a See also reference.

Reply:
Thanks for finding information on probing. We should consider moving all of interop to the .NET Guide. That can be done later, once the revisions and migrations have been completed.